homeassistant/components/qwikswitch.py

```diff
@@ -0,0 +1,180 @@
+    be used in a class that extends both QSToggleEntity *and* ToggleEntity.
+
+    Implemented:
+     - QSLight extends QSToggleEntity and Light[2] (ToggleEntity[1])
+     - QSSwitch extends QSToggleEntity and SwitchDevice[3] (ToggleEntity[1])
+
+    [1] https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity.py  # NOQA
+    [2] https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/light/__init__.py  # NOQA
+    [3] https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/__init__.py  # NOQA
+    """
+
+    def __init__(self, qsitem, qsusb):
+        """Initialize the light."""
+        self._id = qsitem['id']
+        self._name = qsitem['name']
+        self._qsusb = qsusb
+        self._value = qsitem.get('value', 0)
+        self._dim = qsitem['type'] == 'dim'
+
+    @property
+    def brightness(self):
+        """Return the brightness of this light between 0..100."""
+        return self._value if self._dim else None
+
+    # pylint: disable=no-self-use
+    @property
+    def should_poll(self):
+        """State Polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the light."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Check if On (non-zero)."""
+        return self._value > 0
+
+    def update_value(self, value):
+        """Decode QSUSB value & update HA state."""
+        self._value = value
+        # pylint: disable=no-member
+        super().update_ha_state()  # Part of Entity/ToggleEntity
+        return self._value
+
+    # pylint: disable=unused-argument
+    def turn_on(self, **kwargs):
+        """Turn the device on."""
+        from homeassistant.components.light import ATTR_BRIGHTNESS
+        if ATTR_BRIGHTNESS in kwargs:
+            self._value = kwargs[ATTR_BRIGHTNESS]
+        else:
+            self._value = 100
+        return self._qsusb.set(self._id, self._value)
+
+    # pylint: disable=unused-argument
+    def turn_off(self, **kwargs):
+        """Turn the device off."""
+        return self._qsusb.set(self._id, 0)
+
+
+# pylint: disable=too-many-locals
+def setup(hass, config):
+    """Setup the QSUSB component."""
+    from pyqwikswitch import QSUsb
+    from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+    from homeassistant.components.light import Light
+    from homeassistant.components.switch import SwitchDevice
+    from homeassistant.components.discovery import discover
+
+    try:
+        url = config[DOMAIN].get('url', 'http://127.0.0.1:2020')
+        qsusb = QSUsb(url, _LOGGER)
+
+        # Ensure qsusb terminates threads correctly
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP,
+                             lambda event: qsusb.stop())
+    except ValueError as val_err:
+        _LOGGER.error(str(val_err))
+        return False
+
+    global ADD_DEVICES  # pylint: disable=global-statement
```

Comment:
This is not good and a hack. Instead, restructure your component in such a way that both light and switch platforms listen to the callback and handle the appropriate events themselves. That way the main component does not have to be aware of the `add_devices` component methods.